designs/0034-jacobian.md

```diff
@@ -0,0 +1,381 @@
+
+## Constraint Block
+
+### Overview
+
+The constraints block was included in Stan 2.XX to provide users with the ability to define custom constraints on parameters, transformed parameters, and data. 
+This block is situated immediately after the `functions` block and before the `data` block in a Stan program. 
+The primary purpose of the `constraints` block is to allow users to define custom constraints that can be used in the same way as built-in constraints.
+
+### Syntax
+
+A constraints block is defined by the keyword `constraints`. 
+Inside of the `constraints` block each constraint is its own block with a given `constraint_name` and the functions needed to define the constraint.
+Inside the `constraints` block, users can define one or more constraint types. Each constraint type is encapsulated within its own set of curly braces and is identified by a unique name.
+
+Each constraint type must specify a set of functions that define the transformation between unconstrained and constrained spaces, validate data, and increment the log of the absolute value of the determinant of the Jacobian. 
+The first argument for all functions defined in a constraint must be the argument of interest to be transformed or validated. These functions include:
+
+- `constrain(TransformType, OtherArguments...) -> ReturnType`: Transforms a real unconstrained variable to a constrained space.
+- `unconstrain(TransformType, OtherArguments...) -> ReturnType`: Transforms a real constrained variable back to the unconstrained space.
+- `validate(TransformType, OtherArguments...) -> ReturnType`: Validates a real variable against the constraint, returning 1 if valid and 0 otherwise.
+
+The pseudocode below shows the structure of a user defined constrained.
+All CamelCase names are to be replaced by the user.
+
+- `ConstraintName`: The name of the constraint which will be called by users from within the `<>` of Stan's [data types](https://mc-stan.org/docs/reference-manual/overview-of-data-types.html).
+- `ReturnType`: A Stan data type that is the return type of the function
+- `UnconstrainType`: A Stan data type that is the variable to be transformed or validated.
+- `OtherArgs` The rest of the Stan data types that are passed to the constraint during construction of the data type that is being constrained.
+
+
+```stan
+constraints {
+ ConstraintName {
+   // Transform from unconstrained to constrained space
+   ReturnType constrain(UnconstrainType, OtherArgs...) {
+     // ... transformation logic ...
+     // Increment Jacobian
+     jacobian += ...;
+     return ...;
+   }
+
+   // Transform from constrained to unconstrained space
+   ReturnType unconstrain(UnconstrainType, OtherArgs...) {
+     // ... inverse transformation logic ...
+     return ...;
+   }
+
+   // Validate data
+   int validate(UnconstrainType, OtherArgs...) {
+     // ... validation logic ...
+     return ...; // typically returns a boolean value
```

Comment:
We could also let the user `reject()` here - some mention of what exceptions would do in each of these is probably needed.

Reply:
Also, we should state how the `int` is interpreted - presumably 0 is a failure and anything else is fine?